File: algorithms/js/twoPointers/twoSums.js
// Given an array of integers nums and an integer target, return the indices i and j such that nums[i] + nums[j] == target and i != j.

// You may assume that every input has exactly one pair of indices i and j that satisfy the condition.

// Return the answer with the smaller index first.

const twoSumWithPointers = (nums, target) => {
    const numTuples = nums.map((n, i) => [n, i]);
    
    numTuples.sort((a, b) => a[0] - b[0]);

    let left = 0;
    let right = nums.length - 1;

    while (left < right) {
        let sum = numTuples[left][0] + numTuples[right][0];

        if (sum === target) {
            return [numTuples[left][1], numTuples[right][1]];
        }

        if (sum < target) {
            left++;
        } else {
            right--;
        }
    }

    return [];
}

File: algorithms/py/arrays/hashing/twoSums.py
# Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.
# You may assume that each input would have exactly one solution, and you may not use the same element twice.
# You can return the answer in any order.

# Pattern:
#     This is a complement and hashmap problem

# Approach:
#     I iterate through the array and compute the complement (target - current number).
#     If the complement exists in the hashmap, I return the stored index and the current index.
#     Otherwise, I store the current number and its index in the hashmap.

# Time & Space complexity:
#     Time complexity is O(n)
#     Space complexity is O(n)

# Edge Case(s):
#     Handle duplicate numbers at different index

class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        hashmap = {}

        for i, num in enumerate(nums):
            diff = target - num

            if diff in hashmap:
                return [hashmap[diff], i]

            hashmap[num] = i

        return []